### PR TITLE
Feature: Open Contest on elitebot.dev

### PR DESCRIPTION
**Pull request by Obsidianninja11**
## What
Adds a keybind to open a contest on elitebot.dev when clicking with the key held.
Also supports opening other contest pages on elitebot.dev such as upcoming, main page, and your contests.
Allows opening contests in calendar.

I realized after making this that this is a dupe of #915. I think my logic is a bit simpler, though.

## Changelog New Features
+ Added Open On Elite. - Obsidian
    * Allow opening farming contest stats on elitebot.dev by pressing a keybind + mouse click onto a contest item.
    * Works inside the menus Jacob's Farming Contest, Your Contests, and SkyBlock Calendar.

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/config/features/inventory/JacobFarmingContestConfig.java [hunk 1]
```diff
@@ -18,6 +20,11 @@ public class JacobFarmingContestConfig {
     @FeatureToggle
     public boolean realTime = true;
 
+    @Expose
+    @ConfigOption(name = "Open Online", desc = "Opens the contest on §eelitebot.dev§7 when clicking with this key held.")
+    @ConfigEditorKeybind(defaultKey = Keyboard.KEY_NONE)
+    public int openOnline = Keyboard.KEY_NONE;
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/inventory/JacobFarmingContestConfig.java]: rename the toggle to "open elitebot" instead

File: src/main/java/at/hannibal2/skyhanni/features/garden/GardenNextJacobContest.kt [hunk 2]
```diff
@@ -5,29 +5,17 @@ import at.hannibal2.skyhanni.config.ConfigFileType
...
 import at.hannibal2.skyhanni.config.enums.OutsideSbFeature
 import at.hannibal2.skyhanni.config.features.garden.NextJacobContestConfig.ShareContestsEntry
-import at.hannibal2.skyhanni.events.ConfigLoadEvent
-import at.hannibal2.skyhanni.events.DebugDataCollectEvent
-import at.hannibal2.skyhanni.events.GuiRenderEvent
-import at.hannibal2.skyhanni.events.InventoryCloseEvent
-import at.hannibal2.skyhanni.events.InventoryFullyOpenedEvent
-import at.hannibal2.skyhanni.events.LorenzTickEvent
-import at.hannibal2.skyhanni.events.TabListUpdateEvent
+import at.hannibal2.skyhanni.events.*
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/garden/GardenNextJacobContest.kt]: do not use wildcard imports. we have a global setting file to disable this behaviour in intellij.

File: src/main/java/at/hannibal2/skyhanni/features/garden/contest/FarmingContestAPI.kt [hunk 3]
```diff
@@ -1,11 +1,7 @@
 package at.hannibal2.skyhanni.features.garden.contest
 
 import at.hannibal2.skyhanni.data.ScoreboardData
-import at.hannibal2.skyhanni.events.FarmingContestEvent
-import at.hannibal2.skyhanni.events.GuiContainerEvent
-import at.hannibal2.skyhanni.events.InventoryCloseEvent
-import at.hannibal2.skyhanni.events.InventoryFullyOpenedEvent
-import at.hannibal2.skyhanni.events.LorenzTickEvent
+import at.hannibal2.skyhanni.events.*
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/garden/contest/FarmingContestAPI.kt]: same problem here

File: src/main/java/at/hannibal2/skyhanni/features/garden/contest/JacobFarmingContestsInventory.kt [hunk 4]
```diff
@@ -2,27 +2,28 @@ package at.hannibal2.skyhanni.features.garden.contest
...
 import at.hannibal2.skyhanni.SkyHanniMod
 import at.hannibal2.skyhanni.config.ConfigUpdaterMigrator
-import at.hannibal2.skyhanni.events.GuiContainerEvent
-import at.hannibal2.skyhanni.events.GuiRenderItemEvent
-import at.hannibal2.skyhanni.events.InventoryCloseEvent
-import at.hannibal2.skyhanni.events.InventoryUpdatedEvent
-import at.hannibal2.skyhanni.events.LorenzToolTipEvent
-import at.hannibal2.skyhanni.utils.InventoryUtils
+import at.hannibal2.skyhanni.data.HypixelData
+import at.hannibal2.skyhanni.events.*
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/garden/contest/JacobFarmingContestsInventory.kt]: and here

File: src/main/java/at/hannibal2/skyhanni/features/garden/contest/JacobFarmingContestsInventory.kt [hunk 5]
```diff
@@ -73,6 +74,66 @@ class JacobFarmingContestsInventory {
         realTime[slot] = "$dayFormat $startTimeFormat-$endTimeFormat"
     }
 
+    @SubscribeEvent
+    fun onSlotClick(event: SlotClickEvent) {
+        if (!config.openOnline.isKeyHeld()) return
+        if (!LorenzUtils.inSkyBlock) return
+        val invName = InventoryUtils.openInventoryName()
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/garden/contest/JacobFarmingContestsInventory.kt]: dont use abbreviations (inventoryName)

File: src/main/java/at/hannibal2/skyhanni/features/garden/contest/JacobFarmingContestsInventory.kt [hunk 6]
```diff
@@ -73,6 +74,66 @@ class JacobFarmingContestsInventory {
...
+        val name = event.slot.stack.name
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/garden/contest/JacobFarmingContestsInventory.kt]: rename to itemName to readability

File: src/main/java/at/hannibal2/skyhanni/features/garden/contest/JacobFarmingContestsInventory.kt [hunk 7]
```diff
@@ -73,6 +74,66 @@ class JacobFarmingContestsInventory {
...
+
+        when (invName) {
+            "Your Contests" -> {
+                FarmingContestAPI.timePattern.matchMatcher(name) {
+                    val date = getDate(group("year"), group("month"), group("day"))
+                    OSUtils.openBrowser("https://elitebot.dev/contests/$date")
+                    ChatUtils.chat("Opening contest in elitebot.dev")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/garden/contest/JacobFarmingContestsInventory.kt]: Those three lines are duplicated a bit below. Create a function `openContestWebsite(year, month, day)` instead and use it twice

File: src/main/java/at/hannibal2/skyhanni/features/garden/contest/JacobFarmingContestsInventory.kt [hunk 8]
```diff
@@ -73,6 +74,66 @@ class JacobFarmingContestsInventory {
         realTime[slot] = "$dayFormat $startTimeFormat-$endTimeFormat"
     }
 
+    @SubscribeEvent
+    fun onSlotClick(event: SlotClickEvent) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/garden/contest/JacobFarmingContestsInventory.kt]: This function is very long. Try to extract it into multiple smaller functions.

File: src/main/java/at/hannibal2/skyhanni/config/features/inventory/JacobFarmingContestConfig.java [hunk 9]
```diff
@@ -18,6 +20,11 @@ public class JacobFarmingContestConfig {
     @FeatureToggle
     public boolean realTime = true;
 
+    @Expose
+    @ConfigOption(name = "Open Online", desc = "Opens the contest on §eelitebot.dev§7 when clicking with this key held.")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/inventory/JacobFarmingContestConfig.java]: Add the info that this works both in jacobs menu and calendar.

**Review by hannibal002 — CHANGES_REQUESTED** [review]
Works for me.

Can we change it to open on key click alone, not key press + mouse click?
Also maybe show the information "press <keybind> to open this event in elite farming" in the item tooltip.

**Comment by Obsidianninja11** [comment]
> Can we change it to open on key click alone, not key press + mouse click?

Maybe add a toggle?
I like using ctrl+click for this

**Comment by Obsidianninja11** [comment]
> Can we change it to open on key click alone, not key press + mouse click?

I didn't do this.
I don't like it that way.

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**